In [1]:
import pandas as pd
import numpy as np
from IPython.display import Image

In [8]:
anime = pd.read_csv("./Data/AnimeList.csv")
anime.head()

anime_id                title              title_english title_japanese  \
0     11013        Inu x Boku SS  Inu X Boku Secret Service         妖狐×僕SS   
1      2104     Seto no Hanayome      My Bride is a Mermaid          瀬戸の花嫁   
2      5262   Shugo Chara!! Doki         Shugo Chara!! Doki    しゅごキャラ！！どきっ   
3       721        Princess Tutu              Princess Tutu      プリンセスチュチュ   
4     12365  Bakuman. 3rd Season                   Bakuman.          バクマン。   

                                  title_synonyms  \
0                                Youko x Boku SS   
1                           The Inland Sea Bride   
2  Shugo Chara Ninenme, Shugo Chara! Second Year   
3                                            NaN   
4                               Bakuman Season 3   

                                           image_url type    source  episodes  \
0  https://myanimelist.cdn-dena.com/images/anime/...   TV     Manga        12   
1  https://myanimelist.cdn-dena.com/images/anime/...   TV     Manga        26   
2  https://myanimelist.cdn-dena.com/images/anime/...   TV     Manga        51   
3  https://myanimelist.cdn-dena.com/images/anime/...   TV  Original        38   
4  https://myanimelist.cdn-dena.com/images/anime/...   TV     Manga        25   

            status  ...                                         background  \
0  Finished Airing  ...  Inu x Boku SS was licensed by Sentai Filmworks...   
1  Finished Airing  ...                                                NaN   
2  Finished Airing  ...                                                NaN   
3  Finished Airing  ...  Princess Tutu aired in two parts. The first pa...   
4  Finished Airing  ...                                                NaN   

     premiered           broadcast  \
0  Winter 2012  Fridays at Unknown   
1  Spring 2007             Unknown   
2    Fall 2008             Unknown   
3  Summer 2002  Fridays at Unknown   
4    Fall 2012             Unknown   

                                             related  \
0  {'Adaptation': [{'mal_id': 17207, 'type': 'man...   
1  {'Adaptation': [{'mal_id': 759, 'type': 'manga...   
2  {'Adaptation': [{'mal_id': 101, 'type': 'manga...   
3  {'Adaptation': [{'mal_id': 1581, 'type': 'mang...   
4  {'Adaptation': [{'mal_id': 9711, 'type': 'mang...   

                                            producer          licensor  \
0  Aniplex, Square Enix, Mainichi Broadcasting Sy...  Sentai Filmworks   
1                  TV Tokyo, AIC, Square Enix, Sotsu        Funimation   
2                                    TV Tokyo, Sotsu               NaN   
3                 Memory-Tech, GANSIS, Marvelous AQL         ADV Films   
4                                      NHK, Shueisha               NaN   

             studio                                     genre  \
0  David Production    Comedy, Supernatural, Romance, Shounen   
1             Gonzo  Comedy, Parody, Romance, School, Shounen   
2         Satelight             Comedy, Magic, School, Shoujo   
3    Hal Film Maker    Comedy, Drama, Magic, Romance, Fantasy   
4         J.C.Staff           Comedy, Drama, Romance, Shounen   

                                       opening_theme  \
0                              ['"Nirvana" by MUCC']   
1                 ['"Romantic summer" by SUN&LUNAR']   
2  ['#1: "Minna no Tamago (みんなのたまご)" by Shugo Cha...   
3             ['"Morning Grace" by Ritsuko Okazaki']   
4  ['#1: "Moshimo no Hanashi (もしもの話)" by nano.RIP...   

                                        ending_theme  
0  ['#1: "Nirvana" by MUCC (eps 1, 11-12)', '#2: ...  
1  ['#1: "Ashita e no Hikari (明日への光)" by Asuka Hi...  
2  ['#1: "Rottara Rottara (ロッタラ ロッタラ)" by Buono! ...  
3  ['"Watashi No Ai Wa Chiisaikeredo" by Ritsuko ...  
4  ['#1: "Pride on Everyday" by Sphere (eps 1-13)...  

[5 rows x 31 columns]

In [3]:
anime.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14478 entries, 0 to 14477
Data columns (total 31 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   anime_id        14478 non-null  int64  
 1   title           14478 non-null  object 
 2   title_english   5724 non-null   object 
 3   title_japanese  14443 non-null  object 
 4   title_synonyms  8937 non-null   object 
 5   image_url       14382 non-null  object 
 6   type            14478 non-null  object 
 7   source          14478 non-null  object 
 8   episodes        14478 non-null  int64  
 9   status          14478 non-null  object 
 10  airing          14478 non-null  bool   
 11  aired_string    14478 non-null  object 
 12  aired           14478 non-null  object 
 13  duration        14478 non-null  object 
 14  rating          14478 non-null  object 
 15  score           14478 non-null  float64
 16  scored_by       14478 non-null  int64  
 17  rank            12904 non-null 

In [4]:
anime.columns

Index(['anime_id', 'title', 'title_english', 'title_japanese',
       'title_synonyms', 'image_url', 'type', 'source', 'episodes', 'status',
       'airing', 'aired_string', 'aired', 'duration', 'rating', 'score',
       'scored_by', 'rank', 'popularity', 'members', 'favorites', 'background',
       'premiered', 'broadcast', 'related', 'producer', 'licensor', 'studio',
       'genre', 'opening_theme', 'ending_theme'],
      dtype='object')

Well, that's a lot! We should drop some columns

I have already think of dropping the titles, image's URL, Opening and Ending theme since they are redundant and do not provide anything new about the animes. But what are the scored_by, popularity,etc?
Let's learn about those columns directly from MyAnimeList and the data itself. 

Well, let's try to explain each column, one by one.

First of all, the `anime_id` column is the unique ID for every anime on the list. It is a primary key to distinguish every anime to each other.

The `title` column is the anime's names. Similarly, the `title_english`, `title_japanese`, `title_synonyms` are literaly the same to `title`, but they are in different languages and their synonyms, which implies they are redundant and we can safely drop them.

The `image_url` points to the poster image of the anime. We can drop them since it is a link and it won't add any new information to the anime.

There is the easy part, now we must browse our data for more information. 
 

In [12]:
anime = anime.drop(columns=['title_english', 'title_japanese','title_synonyms','image_url'])
anime.columns

Index(['anime_id', 'title', 'type', 'source', 'episodes', 'status', 'airing',
       'aired_string', 'aired', 'duration', 'rating', 'score', 'scored_by',
       'rank', 'popularity', 'members', 'favorites', 'background', 'premiered',
       'broadcast', 'related', 'producer', 'licensor', 'studio', 'genre',
       'opening_theme', 'ending_theme'],
      dtype='object')

I guess the `type` column show what type that anime is. Is it TV or Movie or something else? Let's look at the database for more information.

In [5]:
anime.type.unique()

array(['TV', 'Movie', 'Music', 'OVA', 'ONA', 'Special', 'Unknown'],
      dtype=object)

Exactly! We can spot that there are 7 types of animes. But there is a "weird" type: `Unknown`. Is this a specific type or is this an abbreviate for the `NaN` value?

According to MyAnimeList Search,

![MAL Search](./Images/Anime_Types.png)

there are only 5 types: `TV`, `Movie`, `OVA`, `ONA`, `Special`. So, what do `Music` and `Unknown` mean?

Let take a look at `Unknown` first, since it is the most "weird" type.

In [6]:
print(anime.type.isnull().any())
print(anime.type.isna().any())
anime[anime['type'] == 'Unknown'].head()

False
False


anime_id                                           title  \
187      35759                              Sidonia no Kishi 3   
511      37651  Fate/Extra: Last Encore - Irusterias Tendouron   
1396     35252                                 Hatena☆Illusion   
1989     36407                                   Kenja no Mago   
2326     34649  Soukyuu no Fafner: Dead Aggressor - The Beyond   

           title_english                      title_japanese  \
187   Knights of Sidonia                           シドニアの騎士 3   
511                  NaN   Fate/EXTRA Last Encore イルステリアス天動説   
1396                 NaN                         はてな☆イリュージョン   
1989                 NaN                                賢者の孫   
2326                 NaN  蒼穹のファフナー Dead Aggressor THE BEYOND   

                                         title_synonyms  \
187                                                 NaN   
511   Fate/Extra: Last Encore - Irusterias Ptolemaic...   
1396                                                NaN   
1989  Philosopher&#039;s Grandson, Magi&#039;s Grandson   
2326                                                NaN   

                                              image_url     type       source  \
187   https://myanimelist.cdn-dena.com/images/anime/...  Unknown        Manga   
511   https://myanimelist.cdn-dena.com/images/anime/...  Unknown         Game   
1396  https://myanimelist.cdn-dena.com/images/anime/...  Unknown  Light novel   
1989  https://myanimelist.cdn-dena.com/images/anime/...  Unknown  Light novel   
2326  https://myanimelist.cdn-dena.com/images/anime/...  Unknown     Original   

      episodes         status  ...  background premiered broadcast  \
187          0  Not yet aired  ...         NaN       NaN       NaN   
511          0  Not yet aired  ...         NaN       NaN       NaN   
1396         0  Not yet aired  ...         NaN       NaN       NaN   
1989         0  Not yet aired  ...         NaN       NaN       NaN   
2326         0  Not yet aired  ...         NaN       NaN       NaN   

                                                related producer  licensor  \
187   {'Adaptation': [{'mal_id': 13723, 'type': 'man...      NaN       NaN   
511   {'Adaptation': [{'mal_id': 32425, 'type': 'man...      NaN       NaN   
1396  {'Adaptation': [{'mal_id': 80851, 'type': 'man...      NaN       NaN   
1989  {'Adaptation': [{'mal_id': 90561, 'type': 'man...      NaN       NaN   
2326  {'Other': [{'mal_id': 75, 'type': 'anime', 'ur...      NaN       NaN   

          studio                                   genre  opening_theme  \
187          NaN    Action, Sci-Fi, Space, Mecha, Seinen             []   
511          NaN                  Action, Magic, Fantasy             []   
1396         NaN    Comedy, Supernatural, Romance, Ecchi             []   
1989         NaN          Action, Comedy, Magic, Fantasy             []   
2326  Xebec Zwei  Action, Military, Sci-Fi, Drama, Mecha             []   

      ending_theme  
187             []  
511             []  
1396            []  
1989            []  
2326            []  

[5 rows x 31 columns]

We can see that all the `Unknown` type animes are not aired yet. Thus, we do not have any information about these animes yet, except their name, their related anime/manga, their sources and their genres. We can consider to keep them, since they will show us the trend of future airing animes.

Now, let take a look at `Music` type animes. 

In [9]:
anime[anime['type'] == 'Music'].shape

(849, 31)

A few Google search show that all the `Music` type "Anime" are all music videos. It means that they are not related to our interest. We can drop all the `Music` type entries.

In [11]:
anime = anime[anime['type'] != 'Music']
anime.shape

(13629, 31)

In [13]:
print("We check the columns of the dataset:")
anime.columns

We check the columns of the dataset:


Index(['anime_id', 'title', 'type', 'source', 'episodes', 'status', 'airing',
       'aired_string', 'aired', 'duration', 'rating', 'score', 'scored_by',
       'rank', 'popularity', 'members', 'favorites', 'background', 'premiered',
       'broadcast', 'related', 'producer', 'licensor', 'studio', 'genre',
       'opening_theme', 'ending_theme'],
      dtype='object')

We can easily know that the `source` of the animes are Game, Manga or something similar to that. Let's check all the available source of the animes:

In [14]:
anime['source'].unique()

array(['Manga', 'Original', 'Light novel', '4-koma manga', 'Novel',
       'Visual novel', 'Unknown', 'Other', 'Game', 'Picture book',
       'Card game', 'Web manga', 'Book', 'Music', 'Radio',
       'Digital manga'], dtype=object)

Easily, we can spot the `Unknown` source and may think that it is associated with the `Unknown` type of animes above. Let check that.

In [18]:
anime[anime['type'] == 'Unknown']['source'].value_counts()

Original        10
Manga            7
Light novel      4
Unknown          3
Game             2
4-koma manga     2
Novel            1
Other            1
Name: source, dtype: int64

It seems that almost all `Unknown` type animes have their own source different than `Unknown`. Therefore, `Unknown` source and `Unknown` type is irrelevant to each other. That raise a question, what does `Unknown` source mean?

First of all, let take a look at some `Unknown` source animes.

In [21]:
anime[anime['source'] == 'Unknown'].shape

(4006, 27)

Wow, there are 4006 `Unknown` source animes! Let look at some of them:

In [22]:
anime[anime['source'] == 'Unknown'].head()

anime_id                                       title type   source  \
54     11547                     Sengoku☆Paradise Kiwami   TV  Unknown   
58      7605                              Parasol Henbei   TV  Unknown   
59      9389          Ginga Hyouryuu Vifam: Kieta 12-nin  OVA  Unknown   
61      9998  Chuuhai Lemon: Love 30s - Ame ni Nurete mo  OVA  Unknown   
64      5588                                    Ko Ko Ro  OVA  Unknown   

    episodes           status  airing                  aired_string  \
54        26  Finished Airing   False   Oct 4, 2011 to Mar 27, 2012   
58       200  Finished Airing   False   Oct 2, 1989 to Jan 28, 1991   
59         1  Finished Airing   False                  Feb 25, 1985   
61         1  Finished Airing   False                  Oct 21, 1985   
64         2  Finished Airing   False  Dec 21, 2001 to Jun 21, 2002   

                                         aired         duration  ...  \
54  {'from': '2011-10-04', 'to': '2012-03-27'}   3 min. per ep.  ...   
58  {'from': '1989-10-02', 'to': '1991-01-28'}  10 min. per ep.  ...   
59  {'from': '1985-02-25', 'to': '1985-02-25'}          57 min.  ...   
61  {'from': '1985-10-21', 'to': '1985-10-21'}          44 min.  ...   
64  {'from': '2001-12-21', 'to': '2002-06-21'}  30 min. per ep.  ...   

   background  premiered  broadcast  \
54        NaN  Fall 2011    Unknown   
58        NaN  Fall 1989    Unknown   
59        NaN        NaN        NaN   
61        NaN        NaN        NaN   
64        NaN        NaN        NaN   

                                              related  \
54                                                 []   
58                                                 []   
59  {'Parent story': [{'mal_id': 2987, 'type': 'an...   
61                                                 []   
64                                                 []   

                             producer  licensor              studio  \
54                   TV Tokyo, G-mode       NaN  Milky Cartoon, LMD   
58                                NaN       NaN                 NaN   
59                                NaN       NaN             Sunrise   
61  Tsuchida Productions, Wonder Kids       NaN                 NaN   
64                                NaN       NaN                 NaN   

                                                genre  \
54  Action, Comedy, Drama, Historical, Romance, Sa...   
58            Adventure, Comedy, Fantasy, Kids, Magic   
59                    Sci-Fi, Adventure, Space, Mecha   
61                     Police, Comedy, Romance, Drama   
64                                             Hentai   

                                        opening_theme  \
54  ['#01: "Return Ranse Dokugan Ryuu (リターン乱世独眼竜)"...   
58                                                 []   
59                                                 []   
61                                                 []   
64                                                 []   

                                         ending_theme  
54  ['"Return Ranse Dokugan Ryuu (リターン乱世独眼竜)" by T...  
58                                                 []  
59                                                 []  
61                                                 []  
64                                                 []  

[5 rows x 27 columns]

It seems that the `Unknown` source animes have a lot of `NaN` value in their attributes. But notice that all five animes above finished airing. Maybe most of `Unknown` source animes are finished?

In [23]:
anime[anime['source'] == 'Unknown']['airing'].value_counts()

False    3988
True       18
Name: airing, dtype: int64

Yeah, that is right. We also notice that, the aired date of these animes are quite old. Maybe